tmt/base.py

```diff
@@ -1380,34 +1380,55 @@ def tests(self, keys=None, names=None, filters=None, conditions=None,
         conditions = (conditions or []) + list(Test._opt('conditions', []))
         links = (links or []) + list(Test._opt('links', []))
         excludes = (excludes or []) + list(Test._opt('exclude', []))
+        # Used in: tmt run test --name NAME, tmt test ls NAME...
+        cmd_line_names = list(Test._opt('names', []))
+        source = Test._opt('source', False)
 
         def name_filter(nodes):
             """ Filter nodes based on names provided on the command line """
-            # Used in: tmt run test --name NAME, tmt test ls NAME...
-            names = list(Test._opt('names', []))
-            if not names:
+            if not cmd_line_names:
                 return nodes
             return [
                 node for node in nodes
-                if any([re.search(name, node.name) for name in names])]
-
-        # First let's build the list of test objects based on keys & names.
-        # If duplicate test names are allowed, match test name/regexp
-        # one-by-one and preserve the order of tests within a plan.
-        if not unique and names:
-            tests = []
-            for name in names:
+                if any([re.search(name, node.name) for name in cmd_line_names])]
+
+        def defined_by_source(obj_sources, query_sources):
+            """ True if object_sources contains some string from query_sources """
+            return bool(query_sources.intersection(set(obj_sources)))
+
+        if source:
```

Comment:
What about moving the source selection branch to a separate method? The `tests()` code is getting a bit cluttered.

Reply:
It could also help with sharing the code for plans and stories filtering. If we want that as well...

Reply:
Yes, we definitely want that ;) Not sure about ls and show commands though.
I'll move source selection out

Reply:
Will use https://github.com/teemtee/fmf/pull/175 once approved and merged